File: 3234-count-the-number-of-substrings-with-dominant-ones/3234-count-the-number-of-substrings-with-dominant-ones.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int ans=0;
        int n=s.length();
vector<int>cummSum(n,0);
cummSum[0]=s[0]-'0';
for(int i=1;i<n;i++){
    cummSum[i]=cummSum[i-1]+(s[i]-'0');
}

for(int i=0;i<n;i++){
for(int j=i;j<n;j++){

int oneCnt= i>0 ?cummSum[j]-cummSum[i-1]:cummSum[j];
int zeroCnt=(j-i+1)-oneCnt;

if((zeroCnt*zeroCnt)>oneCnt){
    int diff=(zeroCnt*zeroCnt)-oneCnt;
    j=j+(diff-1);

}

if((zeroCnt*zeroCnt)<=oneCnt){
ans++;

if((zeroCnt*zeroCnt)<oneCnt){
int diff=sqrt(oneCnt)-zeroCnt;

int nextJ=j+diff;
if(nextJ>=n)
ans=ans+(n-j-1);
else
ans=ans+diff;

j=nextJ;
}

}



}

}

        return ans;
    }
};

File: 3234-count-the-number-of-substrings-with-dominant-ones/3234-count-the-number-of-substrings-with-dominant-ones.java
class Solution {
    public int numberOfSubstrings(String s) {
        int maxZeroNum = (int)Math.sqrt(s.length());
        List<Integer> zeroIndex = new ArrayList<>();
        for (int i = 0; i < s.length(); ++i) {
            if (s.charAt(i) == '0') {
                zeroIndex.add(i);
            }
        }
        int ans = 0;
        for (int i = 1; i <= maxZeroNum; ++i) {
            for (int j = 0; j + i - 1 < zeroIndex.size(); ++j) {
                int leftIndex = zeroIndex.get(j);
                int rightIndex = zeroIndex.get(j + i - 1);
                int beforeLeftIndex = (j == 0) ? (0) : (zeroIndex.get(j - 1) + 1);
                int afterRightIndex = j + i >= zeroIndex.size() ? s.length() : zeroIndex.get(j + i);
                for (int start = beforeLeftIndex; start <= leftIndex; start++) {
                    int oneCnt = rightIndex - start + 1 - i;
                    int rightOne = afterRightIndex - rightIndex - 1;
                    ans += (i * i > oneCnt) ? Math.max(rightOne + oneCnt - i * i + 1, 0) : (rightOne + 1);
                }
            }
        }
        for (int i = 0; i < zeroIndex.size(); ++i) {
            int oneCnt = (i == 0) ? zeroIndex.get(i):  zeroIndex.get(i) - zeroIndex.get(i - 1) - 1;
            ans += oneCnt * (oneCnt + 1) / 2;
        }
        int lastOneCnt = zeroIndex.size() > 0 ? s.length() - zeroIndex.get(zeroIndex.size() - 1) - 1 : s.length();
        ans += lastOneCnt * (lastOneCnt + 1) / 2; 
        return ans;
    }
}